runtime/queries/python/highlights.scm

```diff
@@ -1,10 +1,15 @@
-; Identifier naming conventions
+; Types
 
-((identifier) @constructor
- (#match? @constructor "^[A-Z]"))
+(type (identifier) @type)
+(type (subscript (identifier) @type)) ; only one deep...
```

Comment:
By the way, if anyone knows how to make this work for arbitrarily nested values, please let me know. It currently works for something like `dict[str, int]`, but if it is any deeper `dict[str, list[int]]`, the inner identifiers no longer get treated as @type (falls back to @variable I believe).

Reply:
The tree-sitter query language doesn't currently support specifying arbitrary nesting. https://github.com/tree-sitter/tree-sitter/issues/880

In nvim-treesitter I've seen this be addressed by having some large number (~10) of stanzas for increasing depths but it's not very pretty and I think it can make the query analysis time increase potentially noticeably (and a high query analysis will make helix seem to hang while it loads the parser and queries).

Reply:
I see thanks. I just realised I have to also handle the binary operator `|`. Do you think something like this would be an OK compromise:

```scheme
; Handle subscript and | binary operator nesting
(type (_ (identifier) @type)) ; one deep
(type (_ (_ (identifier) @type))) ; two deep
(type (_ (_ (_ (identifier) @type)))) ; three deep
```

Reply:
Implemented the above in new commit, but 4 levels deep.

Reply:
Do you have some code samples I could play around with for these cases? I have an idea for avoiding nesting but I'm not sure it'll work out.

Reply:
Something artificial, but should be suitable for the task. 

```python
class OtherClass:
    def __init__(self, a: int, b: Any):
        self.a = a
        self.b = b

@dataclass
class MyClass:
    field1: OtherClass
    field2: int | None = 5
    field3: dict[str, list[OtherClass | None | int]] | None = None
    
    def func(self):
        self.field1 = OtherClass(3, "anything else")
        x = list(i for i in range(10))
        y = set(x)
```

Reply:
@paul-scott , I think this is better than my PR. However, something that's really helpful on my PR is the separation of @keyword and @keyword.control. It allows the dark_plus theme to go from:
|Current Helix|VSCode|
| --- | --- |
|![image](https://user-images.githubusercontent.com/31261142/167888297-41f2817a-42a6-4eb5-b749-6de8cf27a360.png)|![image](https://user-images.githubusercontent.com/31261142/167888379-4145b7b0-f95b-438a-84ed-5c2b971fc304.png)|

If this replaces my PR, I'd appreciated it if we could fit that in.

Reply:
Just a couple more thoughts.
It would be nice if we can get "None" the type to render differently than "None" the object:
![image](https://user-images.githubusercontent.com/31261142/167889276-811881b5-b99e-4581-93bf-7d54ed1964a8.png)

For some reason, OtherClass gets different highlighting based on whether it is a call or not:
![image](https://user-images.githubusercontent.com/31261142/167889415-1750dff8-ee8e-4b40-ab67-86a64eab4db5.png)

And if it's possible, it would be nice for type aliases to be rendered the same as type declarations:
![image](https://user-images.githubusercontent.com/31261142/167889518-59945f29-f358-4df9-8404-37b27a521796.png)
I know this particular one is asking for a lot! Python mixes statements and types like nobody's business sometimes.

Reply:
@Zeddicus414 you've raised a couple of important points.

1. I didn't consider type aliases, they won't render correctly
2. New type arguments also won't work `UserId = NewType('UserId', int)`
3. The `datetime.datetime` example will highlight both identifiers as if they are types

1 and 2 work under your proposal.

For 3 I don't think there is really anything that can be done in the general case to highlight *just* the type identifier. The behaviour as is now where each identifier is highlighted is in my opinion favourable to either highlight as types.

I suspect we could combine the behaviours in our two PRs to get the best of both worlds?

I know @archseer quoted https://www.programiz.com/python-programming/methods/built-in/str in your PR about `str` etc being pure builtin functions when being called, but I'm not sure that page is up to date given the changes to typing allowing the use of `list` in place of `typing.List` and the like. If I start `ipython` (3.10) I see `str.__class__` and many other fields / methods defined on it. Same for `list` etc. Compared to a function that can't have this. In this sense I think of `str(_)` as a constructor of a string, just like `Foo(_)` would be a constructor for a user defined class.  In this sense I think it makes sense for `str`, `list` etc to be explicitly singled out as either `@type` or `@type.builtin`.

As for None, I kind of like the behaviour as is. The python documentation explicitly singles out `None` and says: "Note that None as a type hint is a special case and is replaced by type(None)." So in that sense `None` on its own isn't a type.

### Priorities

My understanding is that like the https://tree-sitter.github.io implementation of queries, the first match gets priority reading the file top to bottom. This is the exact opposite of how the neovim queries work. For the "alternations" defining the keywords I suspect their position in the file doesn't matter.

### keyword.control operators

Thanks looks good to me. We might want to merge your PR (if others agree), and then I can look to create a new PR on top of it.

What do others think?

Reply:
> This is the exact opposite of how the neovim queries work.

Yeah all implementations follow this order except Neovim. I think the underlying engine is still the same but this behavior comes from how they interact with the queries.

Reply:
Hi @paul-scott , thank you so much for the thoughts and conversation!

### On datetime.datetime coloring
I agree with you about #3 (datetime.datetime). Highlighting the whole thing is preferable and I don't see a way to separate them either.

### On Merging Efforts
I'm definitely up for merging our two efforts, but I'm not entirely sure how to do that. Do I need to fork your forked repo and then create a PR into yours and then you create a PR into helix? Seems super weird. I've only really used git in a corporate environment where I could just make branches directly.

### On str, int functions
I agree about calling str, int, etc type.builtins. It might make sense to do the same for Union, Optional, and the others from the typing module as well. I'm not sure.

### On coloring of None
I disagree about the usage of `None`. I understand your distinction about `None` being a special alias for `type(None)`, but I doubt that the large majority of python users know about that exception... or that it really matters in practice. Consider this example, which is a pretty common idiom for mutable default arguments:
![image](https://user-images.githubusercontent.com/31261142/167980364-f85fc557-2502-4cac-9fb0-29b86d17580c.png)
I think that the `Optional` is clearly easier to visually distinguish and I think  that having the typed `None` green would make it easier to read as well, although, probably still not as easy as `Optional`. No worries if you remain unconvinced though! This is a minor point. And this screenshot is from VSCode again, so you are clearly in good company! I just personally find it kind of jarring to have a blue None on both sides of the equal sign.

### On treesitter priority
Geesh, I've been all over the treesitter documentation and I couldn't find anything about the priority order! I don't know if I just missed it or if it's not there. Thank you two for the pointer! I've only started learning about treesitter at all a couple days ago.

### On Constructors...
I'm not sure about the `@constructor` usefulness either. Seems weird to me.

### Plan forward
We figure out how to get the best of both worlds and get world class python syntax highlighting in Helix!

Reply:
Perhaps it would be best if we work on your current PR. I can create a PR on your fork, and then if / when you merge it it should show up in the upstream PR. That effort can focus on fixing type highlighting and the additional control keyword stuff you implemented.

I can create  a fresh separate PR for the additional non-type stuff I have here. If I can work out how I might mark this one as a draft.

I'll take a look and let you know if this makes some sense.

### On str, int functions

There is a question of where to draw the line with builtin types / builtins in general. I'd say since Optional and Union are not in the prelude but in a module it would be better to leave them as regular types. The catch-all that regexes against things that start with a capital but that aren't a constant should deal with this file.

### On coloring of None

It should be possible to do as you prefer (in type hints, not type alias) for None, at least to a nesting depth `n`. Note that as it isn't an identifier it requires special treatment so will require extra (perhaps negligible?) overhead to implement. I'm on the fence with this one.

### On Constructors...

Perhaps it is best to label things accurately if we can and leave it up to the themes to work out how to best visualise it. I guess it is easy enough to change a theme if you want to make constructor definitions / calls to look like functions and / or types.

Reply:
@paul-scott , I gave you contribute access to my helix fork. We can just work on branches if you want. It makes the most sense to me to work together and update all the python highlighting issues we can and then create one PR, but I'm good either way.

Reply:
@Zeddicus414 you'll probably see the PR on your fork before this, but just in case letting you know I've implemented all the changes I was interested in.